src/main/scala/chisel3/util/BitSet.scala

```diff
@@ -0,0 +1,120 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.util
+
+trait BitSet { bsf =>
+  val terms: Set[BitPat]
```

Comment:
Also, `Set` isn't ordered is it necessary that we make this a `Set` or could it be a `Seq`?

Reply:
I switched to 
```
protected[chisel3] def terms: Set[BitPat]
```
since I moved BitSet decode API to decoder.

Reply:
> Also, Set isn't ordered, is it necessary that we make this a Set or could it be a Seq?

I'd like to extend `BitPat`  from `Ordered` and then we can sort `Set[BitPat]` when require the order.
I still hesitate making terms a `Seq`, since I doubt should we make `terms` public? what if user wanna extends from `BitSet` to implement their own features?

Reply:
> Do we really want the API of BitSet to expose that it's made up of Set[BitPat]s?

I think so, if user wanna implement their own `BitSet` API(like `AddressSet`), `terms` one must be public. And `terms` is the only API which users can access the raw data in a `BitSet`. So I'd like to make this API public.

> Also, Set isn't ordered, is it necessary that we make this a Set or could it be a Seq?

I think Set is OK since there won't be comparing issue, and I have made the serialize API stable via:
```
  override def toString: String = terms.toSeq.sortBy((t: BitPat) => (t.mask, t.value)).mkString("\n")
```

Reply:
> what if user wanna extends from BitSet to implement their own features?

Why do we want the user to extend this? Almost no APIs in Chisel are extensible so I'm not understanding why we should make this particular one extensible.

Reply:
Yes, I think you are right.
I'll seal it :)